src/debug_tools/stats.rs

```diff
@@ -0,0 +1,376 @@
+//! Simple tooling to extract stats from tries.
+//!
+//! This is particularly useful when comparing a "base" trie against a sub-trie
+//! (hashed out trie) created from it.
+
+use std::fmt::{self, Display};
+
+use num_traits::ToPrimitive;
+
+use crate::partial_trie::{Node, PartialTrie};
+
+#[derive(Debug, Default)]
+pub struct TrieStats {
+    pub name: Option<String>,
+    pub counts: NodeCounts,
+    pub depth_stats: DepthStats,
+}
+
+impl Display for TrieStats {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Trie Stats:")?;
+
+        match self.name.as_ref() {
+            Some(name) => writeln!(f, " ({})", name)?,
+            None => writeln!(f)?,
+        }
+
+        writeln!(f, "Counts:\n{}", self.counts)?;
+        writeln!(f, "Depth stats:\n{}", self.depth_stats)
+    }
+}
+
+impl TrieStats {
+    pub fn compare(&self, other: &Self) -> TrieComparison {
+        TrieComparison {
+            node_comp: self.counts.compare(&other.counts),
+            depth_comp: self.depth_stats.compare(&other.depth_stats),
+        }
+    }
+}
+
+/// Total node counts for a trie.
+#[derive(Debug, Default)]
+pub struct NodeCounts {
+    pub empty: usize,
+    pub hash: usize,
+    pub branch: usize,
+    pub extension: usize,
+    pub leaf: usize,
+}
+
+impl Display for NodeCounts {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let tot_nodes = self.total_nodes();
+
+        Self::write_node_count_stats(f, "Empty", self.empty, tot_nodes)?;
+        Self::write_node_count_stats(f, "Hash", self.hash, tot_nodes)?;
+        Self::write_node_count_stats(f, "Branch", self.branch, tot_nodes)?;
+        Self::write_node_count_stats(f, "Extension", self.extension, tot_nodes)?;
+        Self::write_node_count_stats(f, "Leaf", self.leaf, tot_nodes)
+    }
+}
+
+impl NodeCounts {
+    fn write_node_count_stats(
+        f: &mut fmt::Formatter<'_>,
+        node_t_name: &str,
+        count: usize,
+        tot_count: usize,
+    ) -> fmt::Result {
+        let perc = (count as f32 / tot_count as f32) * 100.0;
+        writeln!(f, "{}: {} ({:.3}%)", node_t_name, count, perc)
```

Comment:
very nit: we probably don't need three digits for the percents
```suggestion
        writeln!(f, "{}: {} ({:.2}%)", node_t_name, count, perc)
```